### PR TITLE
Add step to verify repo existence

### PR DESCRIPTION
**Pull request by brunasilvazup**
### Description

This PR add a verification in `rit add repo` (stdin and prompt) to verify if the formula repo has already been added.

* This verification happens after receiving the url and the version.

> Prompt
image

>Stdin
image

>Other version, don't presents error
image

### How to verify it
Clone this PR and add a repo existent, (commons repo for example).

### Changelog

Add step to verify repo existence

### CODE REVIEW

File: pkg/cmd/add_repo_test.go [hunk 1]
```diff
@@ -34,12 +34,30 @@ func TestAddRepoCmd(t *testing.T) {
...
+		Priority: 2,
+	}
+	repoListerPopulated := repoListerAdderCustomMock{
+		add: func(d formula.Repo) error {
+			return nil
+		},
+		list: func() (formula.Repos, error) {
+			return formula.Repos{*repoTest}, nil
+		},
+	}
```
Comment (henriquemoraeszup) [inline pkg/cmd/add_repo_test.go:52]: can we use testify mocks? Its ok that we duplicate a mock for the same purpose, we can remove instances of `repoListerAdderCustomMock` in the future

**Review by lucasdittrichzup — APPROVED** [review]

File: pkg/cmd/add_repo_test.go (hunk 1 above)
Comment (brunasilvazup) [reply pkg/cmd/add_repo_test.go:52]: Yes, i will change this.

**Review by henriquemoraeszup — DISMISSED** [review]
Nice test refact ❤️

**Comment by henriquemoraeszup** [comment]
Hint, maybe to hit the coverage you need a case in which `exitsRepo` returns false

**Comment by brunasilvazup** [comment]
> exitsRepo

Yes, i did in the last test `" Return error when user add a repo existent "`

File: pkg/cmd/add_repo_test.go [hunk 2]
```diff
@@ -54,157 +103,217 @@ func TestAddRepoCmd(t *testing.T) {
...
-					list: func(name string, items []string) (string, error) {
-						return "Github", nil
-					},
+				repo:               returnOfRepoListerAdder{errAdd: nil, reposList: formula.Repos{}, errList: nil},
+				InputTextValidator: returnWithStringErr{"mocked text", nil},
+				InputPassword:      returnWithStringErr{"s3cr3t", nil},
+				InputURL:           returnWithStringErr{"http://localhost/mocked", nil},
+				InputBool:          returnOfInputBool{true, nil},
+				InputList:          []returnOfInputList{{response: "Github", err: nil}},
+				stdin:              "{\"provider\": \"github\", \"name\": \"repo-name\", \"version\": \"\", \"url\": \"https://url.com/repo\", \"token,omitempty\": \"\", \"priority\": 5, \"isLocal\": false}\n",
```
Comment (victorschumacherzup) [inline pkg/cmd/add_repo_test.go:298]: What you think about put the string between `? So the slashes can be removed

Comment (victorschumacherzup) [reply pkg/cmd/add_repo_test.go:298]: Since it is repeated on 284, a const will be nice

Comment (brunasilvazup) [reply pkg/cmd/add_repo_test.go:298]: Hi, it's a good idea, i will be change that, tks (:

Comment (brunasilvazup) [reply pkg/cmd/add_repo_test.go:298]: But this entry hasn't version and in entry of line 284 has, so i will be change only ` ok?

**Review by victor-schumacher — DISMISSED** [review]

**Review by victorschumacherzup — APPROVED** [review]

File: pkg/cmd/add_repo_test.go [hunk 3]
```diff
@@ -54,157 +103,217 @@ func TestAddRepoCmd(t *testing.T) {
...
+		{
+			name: "Run with success when user add a new commons",
+			fields: fields{
+				repo:               returnOfRepoListerAdder{errAdd: nil, reposList: formula.Repos{formula.Repo{Provider: "Github", Name: formula.RepoCommonsName}}, errList: nil},
+				InputTextValidator: returnWithStringErr{formula.RepoCommonsName.String(), nil},
+				InputPassword:      returnWithStringErr{"s3cr3t", nil},
+				InputURL:           returnWithStringErr{"http://localhost/mocked", nil},
+				InputBool:          returnOfInputBool{true, nil},
+				InputList:          []returnOfInputList{{response: "Github", err: nil}},
+				tutorialStatus:     "disabled",
```
Comment (henriquemoraeszup) [inline pkg/cmd/add_repo_test.go:125]: Hey! O love this approach, I was just thinking, could we somehow have a default return in place and only override the returns we want to change? I feel this way we would end up with less repeated code and on each test, we would explicitly have only the mocked field with a different side effect. What do you think?

File: internal/mocks/mocks.go [hunk 4]
```diff
@@ -42,3 +46,117 @@ func (d *DetailManagerMock) LatestTag(repo formula.Repo) string {
...
+
+	return args.Error(0)
+}
+
+type TutorialFindSetterMock struct {
+	mock.Mock
+}
+
+func (t *TutorialFindSetterMock) Find() (rtutorial.TutorialHolder, error) {
+	args := t.Called()
```
Comment (henriquemoraeszup) [inline internal/mocks/mocks.go:71]: Bruna, I was thinking, we might end up with too many mocks. Also the only other folder in `internal` is `proto`, which we are not using. What do you think of a 1st-level folder `mocks` with different scoped files for them? All from `mocks` package but more like
```
mocks
    input.go (bool, list, int)
    formula.go
    actions.go (add, delete, find)
    cmd.go
```

We might end up with a problem of categorizing mocks properly, so I am not sure this is still the best approach. But I am still in favor of moving the `mocks` folder up one level 😄

File: pkg/cmd/add_repo_test.go (hunk 3 above)
Comment (brunasilvazup) [reply pkg/cmd/add_repo_test.go:125]: Do you say something similar to that?
-> 
https://github.com/ZupIT/ritchie-cli/blob/e09272256bd7f000a8aaeeb0eb82ee0a37d9f988/pkg/formula/input/flag/flag_test.go#L144

Comment (henriquemoraeszup) [reply pkg/cmd/add_repo_test.go:125]: That could definitely work, its sad that golang does not have coalesce operator 😢

File: internal/mocks/mocks.go (hunk 4 above)
Comment (henriquemoraeszup) [reply internal/mocks/mocks.go:71]: sounds good

File: pkg/cmd/add_repo.go [hunk 5]
```diff
@@ -241,6 +255,15 @@ func (ad addRepoCmd) repoNameValidator(text interface{}) error {
 	return nil
 }
 
+func exitsRepo(urlToAdd, versionToAdd string, repos formula.Repos) bool {
```
Comment (henriquemoraeszup) [inline pkg/cmd/add_repo.go:258]: nitpick
```suggestion
func existsRepo(urlToAdd, versionToAdd string, repos formula.Repos) bool {
```

**Review by henriquemoraeszup — DISMISSED** [review]

**Review by henriquemoraeszup — APPROVED** [review]

**Pull request merged**